tools/btmidi-server.c

```diff
@@ -0,0 +1,1032 @@
+			memset(&advertise_cp, 0, sizeof(advertise_cp));
+			advertise_cp.enable = 0x01;
+			memset(&rq, 0, sizeof(rq));
+			rq.ogf = OGF_LE_CTL;
+			rq.ocf = OCF_LE_SET_ADVERTISE_ENABLE;
+			rq.cparam = &advertise_cp;
+			rq.clen = LE_SET_ADVERTISE_ENABLE_CP_SIZE;
+			rq.rparam = &status;
+			rq.rlen = 1;
+			err = hci_send_req(dd, &rq, 1000);
+		}
+	}
+	hci_close_dev(dd);
+	return err;
+}
+
+static int l2cap_le_att_listen_and_accept(bdaddr_t *src, int sec,
+							uint8_t src_type)
+{
+	int sk, nsk;
+	struct sockaddr_l2 srcaddr, addr;
+	socklen_t optlen;
+	struct bt_security btsec;
+	char ba[18];
+	int dev_id;
+	
+	sk = socket(PF_BLUETOOTH, SOCK_SEQPACKET, BTPROTO_L2CAP);
+	if (sk < 0) {
+		perror("Failed to create L2CAP socket");
+		return -1;
+	}
+
+	dev_id = hci_get_route(src);
+	if (dev_id < 0) {
+		errno = ENODEV;
+		fprintf(stderr, "Device ID not found\n");
+		goto fail;
+	}
+
+	if (advertise(dev_id) != 0) {
+		fprintf(stderr, "Failed advertising\n");
+		goto fail;
+	}
+
+	/* Set up source address */
+	memset(&srcaddr, 0, sizeof(srcaddr));
+	srcaddr.l2_family = AF_BLUETOOTH;
+	srcaddr.l2_cid = htobs(ATT_CID);
+	srcaddr.l2_bdaddr_type = src_type;
+	bacpy(&srcaddr.l2_bdaddr, src);
+
+	if (bind(sk, (struct sockaddr *) &srcaddr, sizeof(srcaddr)) < 0) {
+		perror("Failed to bind L2CAP socket");
+		goto fail;
+	}
+
+	/* Set the security level */
+	memset(&btsec, 0, sizeof(btsec));
+	btsec.level = sec;
+	if (setsockopt(sk, SOL_BLUETOOTH, BT_SECURITY, &btsec,
+							sizeof(btsec)) != 0) {
+		fprintf(stderr, "Failed to set L2CAP security level\n");
+		goto fail;
+	}
+
+	if (listen(sk, 10) < 0) {
+		perror("Listening on socket failed");
+		goto fail;
+	}
+
+	if (verbose) {
+		printf("Started listening on ATT channel. Waiting for connections\n");
+	}
+
+	memset(&addr, 0, sizeof(addr));
+	optlen = sizeof(addr);
+	nsk = accept(sk, (struct sockaddr *) &addr, &optlen);
+	if (nsk < 0) {
+		perror("Accept failed");
+		goto fail;
+	}
+
+	if (verbose) {
+		ba2str(&addr.l2_bdaddr, ba);
+		printf("Connect from %s\n", ba);
+	}
+	
+	close(sk);
+
+	return nsk;
+
+fail:
+	close(sk);
+	return -1;
+}
+
+static struct server *server_create(int fd, uint16_t mtu, int client, int port)
+{
+	struct server *server;
+	size_t name_len = strlen(gattName);
+	int err;
+
+	server = new0(struct server, 1);
+	if (!server) {
+		fprintf(stderr, "Failed to allocate memory for server\n");
+		return NULL;
+	}
+
+	/* Init values */
+	server->io = NULL;
+	server->seq_port_id = -1;
+	server->seq_handle = NULL;
+
+	server->att = bt_att_new(fd, false);
+	if (!server->att) {
+		fprintf(stderr, "Failed to initialze ATT transport layer\n");
+		goto fail;
+	}
+
+	if (!bt_att_set_close_on_unref(server->att, true)) {
+		fprintf(stderr, "Failed to set up ATT transport layer\n");
+		goto fail;
+	}
+
+	if (!bt_att_register_disconnect(server->att, att_disconnect_cb, server, NULL)) {
+		fprintf(stderr, "Failed to set ATT disconnect handler\n");
+		goto fail;
+	}
+
+#if ENABLE_EXTRA_CHARACTERISTICS
```

Comment:
Yes. This will definitely recommended for us to do. It will help the Master device to setup its link, but we need to do the same on our local link as well.